### DISCUSSION TITLE
instance segmentation parameters

### DISCUSSION THREAD

**Discussion by ktyssowski** [neuropoly/axondeepseg#766] [Help needed]
Axondeepseg is working great for me for separating myelin from axon, but I think the instance segmentation that happens when calculating morphometrics could be better -- in particular, it seems that for axons near thickly myelinated axons, the line between axons errs towards giving the less myelinated axons more area. I'm wondering if there is any way to tweak the instance segmentation process.

Thanks!

**Comment by hermancollin** [comment]
Hi @ktyssowski. There is indeed a few tricks one can use to enhance his instance segmentation. Would you mind posting an example of what you are referring to so we can take a look together?

**Comment by ktyssowski** [comment]
yes! here's my instance map with a few problem areas highlighted, along with the corresponding axon myelin index.

![Kelsey_7181_NB1_cervical001_instance-map](https://github.com/axondeepseg/axondeepseg/assets/8950813/87556407-3baf-49d6-98c5-89ab1d4f9821)
![Kelsey_7181_NB1_cervical001_axonmyelin_index](https://github.com/axondeepseg/axondeepseg/assets/8950813/25061e89-fe8f-4d66-85af-d1fce1ebb705)

ETA: the original image
![NB1_001](https://github.com/axondeepseg/axondeepseg/assets/8950813/a93fc8dd-ce3a-4fa2-bdfd-d1becd27817c)

**Reply by mathieuboudreau** [reply]
One quick suggestion I could make is that, for instance segmentation objects that you've identified as having incorrect boundaries (because the masks will have touching myelin), you could go in with the napari tool and "draw" a small line separating the myelin using the background value ("color" or black & white value, should be 0) from the two objects. This is semi-manual, but would quickly resolve this issue if you don't have **too** many touching axons.

**Reply by ktyssowski** [reply]
Thanks! I've tried something similar (using a 1px eraser instead of drawing a line), and it does work well, but it takes a lot of time, so I was hoping for a less manual solution

if i could improve the non-manual segmentation a bit, i could clean up this way, though

**Reply by hermancollin** [reply]
Hi @ktyssowski . Could you send me the original segmentation masks? Or alternatively which model/pixel size did you use?

I would like to try a postprocessing method I researched a year ago.

**Comment by ktyssowski** [comment]
Sure! Here's the segmentation png
![Kelsey_7181_NB1_cervical001_seg-axonmyelin](https://github.com/axondeepseg/axondeepseg/assets/8950813/e0637c82-415f-474a-9a34-5c95951a6047)

I also used the TEM model, pixel size .003

**Reply by hermancollin** [reply]
@ktyssowski Some time ago, I trained a model to predict the touching areas in the myelin. This model takes as input an axon/myelin mask like the one you provided and outputs a binary mask of the overlapping areas between consecutive fibers. Here is what it predicts for your mask:
![sub-nyuMouse09_sample-0001_TEM_class-0_pred](https://github.com/axondeepseg/axondeepseg/assets/83031821/e5662389-5b97-4644-b360-2b28d86e44e3)
By substracting this mask from the original axonmyelin mask, we get this postprocessed mask:
![user_ktyssowski_mask_postprocessed](https://github.com/axondeepseg/axondeepseg/assets/83031821/b20f4ff1-5f50-431c-9ac3-213b7419d2c3)
It's not perfect but maybe this could help you. Here is the resulting colorized instance map:
![user_ktyssowski_instance-map](https://github.com/axondeepseg/axondeepseg/assets/83031821/ce26692b-4639-427e-82d9-0e2d38e6e41c)
Some very minor corrections to the masks could also go a long way. This model was never publicly released but we could add it in the software as an experimental feature if you are interested.

**Reply by ktyssowski** [reply]
ooo! that looks great! i'd definitely try it if it were and experimental feature!

**Reply by ktyssowski** [reply]
hi @hermancollin -- was this ever added as an experimental feature? just want to make sure i didn't miss it! thanks!

**Reply by hermancollin** [reply]
Hi @ktyssowski . Sorry if I was unclear - no this feature is not currently available. I have no timeline for when this could be included in the main software. In the meantime, maybe you could send us an email at axondeepseg@googlegroups.com and I'll send you the model. Feel free to briefly describe your research project, we are always interested to hear about it.

**Reply by ktyssowski** [reply]
ah i see! thanks! just emailed
